3scaleAdapter/pkg/threescale/proxy_conf.go

```diff
@@ -0,0 +1,108 @@
+package threescale
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	sysC "github.com/3scale/3scale-porta-go-client/client"
+	"github.com/3scale/istio-integration/3scaleAdapter/config"
+
+	"istio.io/istio/pkg/log"
+)
+
+// DefaultCacheTTL - Default time to wait before purging expired items from the cache
+const DefaultCacheTTL = time.Duration(time.Minute * 1)
+const cacheKey = "%s-%s"
+
+type proxyStore struct {
+	element sysC.ProxyConfigElement
+	setAt   time.Time
+}
+
+// proxyConfigController offers a way to cache proxy configurations from
+// 3scale system.
+type proxyConfigController struct {
+	limit int
+	ttl   time.Duration
+	cache map[string]proxyStore
+	mutex sync.RWMutex
+}
+
+func newProxyConfigController(cacheTTL time.Duration, limit int) *proxyConfigController {
+	pcc := &proxyConfigController{
+		limit: limit,
+		ttl:   cacheTTL,
+		cache: make(map[string]proxyStore, limit),
+	}
+	go pcc.flushCache(cacheTTL)
+	return pcc
+}
+
+func (pc *proxyConfigController) get(cfg *config.Params, c *sysC.ThreeScaleClient) (sysC.ProxyConfigElement, error) {
+	var conf sysC.ProxyConfigElement
+	var err error
+
+	cacheKey := getCacheKeyFromCfg(cfg)
+
+	pc.mutex.RLock()
+	e, ok := pc.cache[cacheKey]
+	pc.mutex.RUnlock()
+
+	if !ok {
+		conf, err = pc.getFromRemote(cfg, c)
+	} else {
+		log.Debugf("proxy config fetched from cache for service id %s", cfg.ServiceId)
+		conf = e.element
+	}
+
+	return conf, err
+}
+
+func (pc *proxyConfigController) getFromRemote(cfg *config.Params, c *sysC.ThreeScaleClient) (sysC.ProxyConfigElement, error) {
+	log.Debugf("proxy config for service id %s not cached - fetching from 3scale", cfg.ServiceId)
+	pce, err := c.GetLatestProxyConfig(cfg.AccessToken, cfg.ServiceId, "production")
+	if err == nil {
+		go pc.set(cfg, pce)
+	}
+	return pce, err
+}
+
+func (pc *proxyConfigController) set(cfg *config.Params, e sysC.ProxyConfigElement) {
+	cacheKey := getCacheKeyFromCfg(cfg)
+
+	pc.mutex.Lock()
+	defer pc.mutex.Unlock()
+	if len(pc.cache) < pc.limit {
+		pc.cache[cacheKey] = proxyStore{
+			setAt:   time.Now(),
+			element: e,
+		}
+	}
+}
+
+func getCacheKeyFromCfg(cfg *config.Params) string {
+	return fmt.Sprintf(cacheKey, cfg.SystemUrl, cfg.ServiceId)
+}
+
+func (pc *proxyConfigController) flushCache(t time.Duration) {
+	for {
+		<-time.After(t)
+		now := time.Now()
+
+		pc.mutex.Lock()
+		for k, v := range pc.cache {
+			if now.Sub(v.setAt) >= pc.ttl {
```

Comment:
We can speed this up by avoiding this `Sub` operation here - we can either store an `expiresAt` timestamp in each entry and compare against `now` (placing the operation to compute the expiration timestamp at insertion time), or reuse the `setAt` and computing `maxAge := now.Sub(pc.ttl)` outside the loop and comparing each `setAt` field against `maxAge`.